### PR TITLE
Lazy imports

### PR DESCRIPTION
**Pull request by arthurevans**
Iterating on @stramel's lazy-imports doc. 

cc: @rictic @katejeffreys

### CODE REVIEW

File: app/2.0/docs/glossary.md [hunk 1]
```diff
@@ -94,13 +94,18 @@ This tool has been replaced by [Polymer Analyzer](#polymer-analyzer). It is a st
...
 ### lazy loading
 Lazy loading takes advantage of routing to load elements only when they are required for render. 
 Used correctly, lazy loading can improve performance.
 
-For an example implementation of lazily-loaded elements, see the [Shop 
-app](https://github.com/Polymer/shop). In the Shop app, views are elements which are instantiated 
-only when they are routed to. 
+For an example implementation of lazily-loaded elements, see the [Polymer Starter
+Kit](https://github.com/PolymerElements/polymer-starter-kit). In the starter kit,
+views are elements which are instantiated only when they are routed to.
```
Comment (abdonrd) [inline app/2.0/docs/glossary.md]: Then we should review and merge https://github.com/PolymerElements/polymer-starter-kit/pull/998.

File: app/2.0/docs/glossary.md [hunk 2]
```diff
@@ -94,13 +94,18 @@ This tool has been replaced by [Polymer Analyzer](#polymer-analyzer). It is a st
 framework for Web Components. 
 
 ## L
+### lazy-imports
```
Comment (katejeffreys) [inline app/2.0/docs/glossary.md]: I suggest 

`### lazy imports`

instead of 

`### lazy-imports`

File: app/2.0/docs/glossary.md [hunk 3]
```diff
@@ -94,13 +94,18 @@ This tool has been replaced by [Polymer Analyzer](#polymer-analyzer). It is a st
...
+Lazy Imports are a declarative way to describe the resources that you will import lazily, and
```
Comment (katejeffreys) [inline app/2.0/docs/glossary.md]: This paragraph sounds a bit tautological to me :) I suggest rewording. Maybe something like this, edited for technical accuracy ;) :

"Lazy imports are a declarative way to describe the resources to be loaded lazily in an application (see #lazy-loading). Describing lazily-loaded resources declaratively means they can be linted, and that sharded bundling can be performed without configuration."

What is sharded bundling? We don't mention it in the glossary and idk what it means. I feel that we should define it in the glossary if we use it in another definition. 

Also not sure what it means to be able to do linting and sharded bundling without configuration. What config is needed if the resources are described imperatively? I am confuse :/

Not sure if others would share my confusion, however :)

File: app/2.0/start/toolbox/create-a-page.md [hunk 4]
```diff
@@ -97,29 +97,61 @@ you need to add it to your app's HTML.
...
+        <!-- ... -->
+    ```
+
+1.  Typically when adding a new custom element for the first time, you'd
     want to add an HTML import to ensure the component definition has been
-    loaded.  However, this app template is already set up to lazy-load top
-    level views on-demand based on the route, so in this case you don't need
-    to add an import for your new `<my-new-view>` element.
+    loaded.  However, this app template is already set up to lazy-load views
+    on-demand based on the route, so in this case you need to add a
```
Comment (katejeffreys) [inline app/2.0/start/toolbox/create-a-page.md]: `on-demand` --> `on demand`

File: app/2.0/toolbox/lazy-imports.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+<!-- toc -->
+
+Normal HTML Imports are eager, meaning that they are loaded and evaluated in
+order first, before any code that follows. You can get a large performance
+improvement by lazily loading code at runtime, so that you only load the
+minimal amount of code needed to display the current view. This is a key piece
+of [the PRPL pattern](/2.0/toolbox/prpl).
+
+
+Lazy imports are an optional method for adding declarative lazy-loaded imports. The Polymer CLI
```
Comment (katejeffreys) [inline app/2.0/toolbox/lazy-imports.md]: I suggest deciding on "lazy-loaded" or "lazily-loaded" and sticking to it throughout :)

File: app/2.0/toolbox/lazy-imports.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+can use lazy imports to provide better analysis—for example, providing appropriate lint errors for
+lazy-loaded elements.
+
+## Using lazy imports
+
+The lazy imports mechanism is already built in to the latest version of Polymer starter kit
```
Comment (katejeffreys) [inline app/2.0/toolbox/lazy-imports.md]: "latest version of Polymer starter kit" --> "latest version of the Polymer starter kit"

File: app/2.0/toolbox/lazy-imports.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        ...
+      </template>
+      ...
+    </dom-module>
+    ```
+
+    Note that each `lazy-import` has a `group` attribute. You'll use this `group` name to load
+    the import. Multiple lazy imports can have the same `group` name.
+
+4.  Add the mixin or behavior to the element definition:
```
Comment (katejeffreys) [inline app/2.0/toolbox/lazy-imports.md]: This is in the 2.0 section of the docs, so IMO we should remove the 1.0 syntax.

File: app/2.0/toolbox/prpl.md [hunk 8]
```diff
@@ -68,32 +68,45 @@ When you generate an App Toolbox project using Polymer CLI, the new project cont
 
 The shell is responsible for routing and usually includes the main navigation UI for the app.
 
-The app should call `importHref` to lazy-load fragments as they're required. For example, when the
-user changes to a new route, it imports the fragment(s) associated with that route. This may
-initiate a new request to the server, or simply load the resource from the cache.
+The app should lazy load code at runtime as required to get a large performance
```
Comment (katejeffreys) [inline app/2.0/toolbox/prpl.md]: "The app should lazy load code at runtime" --> "The app should lazy-load code at runtime as required to get a large performance" or maybe "The app should lazily load code at runtime as required to get a large performance"

File: app/2.0/toolbox/prpl.md [hunk 9]
```diff
@@ -68,32 +68,45 @@ When you generate an App Toolbox project using Polymer CLI, the new project cont
...
-    this._importFailedCallback.bind(this), /* for example, display 404 page */
-    true); /* make import async */
+class MyApp extends Polymer.LazyImportsMixin(Polymer.Element) {
+
+    // ...
+
+    _pageChanged(page) {
+        // Import the assets with the associated group
+        this.importLazyGroup(page)
+            .catch(this._importFailedCallback.bind(this)); // Handle error, ie. Display 404 page
```
Comment (katejeffreys) [inline app/2.0/toolbox/prpl.md]: "ie." --> "e.g."

File: app/2.0/toolbox/prpl.md [hunk 10]
```diff
@@ -68,32 +68,45 @@ When you generate an App Toolbox project using Polymer CLI, the new project cont
...
-    true);
+behaviors: [Polymer.LazyImportsBehavior],
+
+// ...
+
+_pageChanged: function(page) {
+    // Import the assets with the associated group
+    this.importLazyGroup(page)
+            .catch(this._importFailedCallback.bind(this)); // Handle error, ie. Display 404 page
+}
```
Comment (katejeffreys) [inline app/2.0/toolbox/prpl.md]: "ie." --> "e.g."

**Review by ghost — APPROVED** [review]

File: app/2.0/docs/glossary.md (hunk 1 above)
Comment (stramel) [inline app/2.0/docs/glossary.md]: Whoops, didn't notice this was on @arthurevans's branch :sweat_smile:

Comment (arthurevans) [inline app/2.0/docs/glossary.md]: @abdonrd Yeah, I was trying to get this ready so that we could roll it out whenever that got merged into the PSK.  (But we could also remove the sentence that says that the latest version of the PSK uses lazy imports. Let me know.)

Comment (abdonrd) [inline app/2.0/docs/glossary.md]: It would be great to have it merged into the PSK. 🙂

**Comment by katejeffreys** [comment]
@stramel Can you pls comment with a statement that you're OK with your commits being included in this PR?

I can go ahead and make the changes I suggested if nobody objects. 

Then this should be mergeable AFAIK. @arthurevans ?

**Comment by stramel** [comment]
@katejeffreys  I am okay with my commits being utilized in this or any PR :smile: 

The only concern I have with merging is that the polymer-starter-kit doesn't actually implement them yet, pending https://github.com/PolymerElements/polymer-starter-kit/pull/998.

There are a lot of things outstanding in the lazy-imports side that need review as well but nothing that should hold up this PR.

**Comment by abdonrd** [comment]
@arthurevans old PR; closing?

**Comment by stramel** [comment]
@arthurevans 	Should we go ahead and close this PR?

**Comment by arthurevans** [comment]
@stramel Yep, I think we should. Thanks for all your work on this.
